shap/utils/_clustering.py

```diff
@@ -228,37 +253,31 @@ def hclust(
 
     # build the distance matrix
     if metric == "xgboost_distances_r2":
-        dist_full = xgboost_distances_r2(X, y, random_state=random_state)
+        dist_full: np.ndarray = xgboost_distances_r2(X_arr, y, random_state=random_state)
 
         # build a condensed upper triangular version by taking the max distance from either direction
-        dist_list = []
-        for i in range(dist_full.shape[0]):
-            for j in range(i + 1, dist_full.shape[1]):
-                if i != j:
-                    if linkage == "single":
-                        dist_list.append(min(dist_full[i, j], dist_full[j, i]))
-                    elif linkage == "complete":
-                        dist_list.append(max(dist_full[i, j], dist_full[j, i]))
-                    elif linkage == "average":
-                        dist_list.append((dist_full[i, j] + dist_full[j, i]) / 2)
-                    else:
-                        raise ValueError("Unsupported linkage type!")
+        dist_list: list[float] = []
```

Comment:
NIT: I would prefer to create the matrix here right away and then fill it line by line later. 

Pro: shape of the matrix is clear as one reads the code, memory is already allocated (though this is neglectable).
Con: we would need to enumerate and another index in the loop.

Reply:
@CloseChoice I'm assuming you mean the creation of the `dist_list`'s array here,

Agree on the memory reallocation point, and yea it probably doesn't matter here. As for the shape:  
I think the number of dimensions is clear enough from the code (i.e., ndim=1); and the length of the dist_list (k**2-k) I would argue isn't that important here since the `dist` array is input into the scipy hclust functions right after.

Hope it's alright with you that I close this